### PR TITLE
NIFIREG-9: Initial Auth Implementation

### PR DESCRIPTION
**Pull request by kevdoran**
This is a rather large PR, including:

- Authentication and authorization enforcement for web API, largely based on NiFi.
- This adds interfaces, framework, and file-based authorizer providers (file access policy provider, file user group provider).
- Authentication of identities is currently based on certificates in two-way SSL (HTTPS). Alternative identity strategies (user&pass, JWT) will be added later building upon the foundation in this commit.

For reviewers, a good place to start is *AuthorizationService.java*. That is called by Resources (web api endpoints) for authorizing requests. It makes use of authorizer providers (access policy provider and user group provider), which are interfaces and implementations heavily based in NiFi.

To run this, see the instructions in included authorizers.xml file. Configure that with your initial admin identity (must be certificate based) and configure your properties file to use your configured authorizers.

This functionality has been verified through manual testing. Documentation on proper configuration (aside what is in the provided in comment ins the included authorizers.xml example file) is in progress. Automated tests cases (unit, integration, etc.) are in progress and will likely be added to the PR during the time it is reviewed, but opening it now to get initial feedback, especially where it defines/modifies key interfaces.

As part of this feature, some changes were made to the RegistryService interface and the providers it utilizes. @bbende please review these changes when you get a change to make sure they are consistent with your design and intentions for the service and provider layers. Thanks!

### CODE REVIEW

**Comment by bbende** [comment]
Reviewing...

File: nifi-registry-security/pom.xml [hunk 1]
```diff
@@ -51,11 +67,25 @@
...
             </plugin>
         </plugins>
     </build>
+
+    <dependencyManagement>
+        <dependencies>
+            <!-- This is here because Spring Security 4.2.3.RELEASE builds against Spring Framwork 4.3.9.RELEASE -->
+            <dependency>
+                <groupId>org.springframework</groupId>
+                <artifactId>spring-framework-bom</artifactId>
```
[the remark this answers is not in the document]
Comment (kevdoran) [reply nifi-registry-security/pom.xml]: Good catch. Yep, will do.

File: nifi-registry-security/src/main/java/org/apache/nifi/registry/authorization/AccessPolicy.java [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * Defines a policy for a set of userIdentifiers to perform a set of actions on a given resource.
+ */
+public class AccessPolicy {
```
[the remark this answers is not in the document]
Comment (kevdoran) [reply nifi-registry-security/src/main/java/org/apache/nifi/registry/authorization/AccessPolicy.java]: It's a good suggestion. I wasn't thrilled with the current organization. Your approach both distinguishes interfaces from implementation and adds extensibility. Will make that change, thanks!

File: nifi-registry-web-api/pom.xml [hunk 3]
```diff
@@ -105,6 +105,11 @@
         </dependency>
         <dependency>
             <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-starter-security</artifactId>
```
[the remark this answers is not in the document]
Comment (kevdoran) [reply nifi-registry-web-api/pom.xml]: Yep.

File: nifi-registry.properties [hunk 4]
```diff
@@ -0,0 +1,47 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
[the remark this answers is not in the document]
Comment (kevdoran) [reply nifi-registry.properties]: yeah this was a test file. will remove

File: nifi-registry-assembly/pom.xml [hunk 5]
```diff
@@ -137,6 +137,8 @@
         <nifi.registry.security.truststorePasswd />
         <nifi.registry.security.needClientAuth />
         <nifi.registry.security.authorized.users>./conf/authorized-users.xml</nifi.registry.security.authorized.users>
+        <nifi.security.user.authorizer>managed-authorizer</nifi.security.user.authorizer>
```
[the remark this answers is not in the document]
Comment (kevdoran) [reply nifi-registry-assembly/pom.xml]: nope. neither is login.identity.provider. will remove unused properties and fix the prefixes to be consistent. good catch!

File: nifi-registry-resources/src/main/resources/conf/nifi-registry.properties [hunk 6]
```diff
@@ -32,6 +32,8 @@ nifi.registry.security.truststoreType=${nifi.registry.security.truststoreType}
 nifi.registry.security.truststorePasswd=${nifi.registry.security.truststorePasswd}
 nifi.registry.security.needClientAuth=${nifi.registry.security.needClientAuth}
 nifi.registry.security.authorized.users=${nifi.registry.security.authorized.users}
+nifi.security.user.authorizer=${nifi.security.user.authorizer}
```
[the remark this answers is not in the document]
Comment (kevdoran) [reply nifi-registry-resources/src/main/resources/conf/nifi-registry.properties]: same as above. removed in the next update

File: nifi-registry-security/pom.xml [hunk 7]
```diff
@@ -34,12 +34,28 @@
...
                 <executions>
                     <execution>
-                        <id>xjc</id>
+                        <id>authorizations</id>
                         <goals>
                             <goal>xjc</goal>
                         </goals>
                         <configuration>
-                            <packageName>org.apache.nifi.registry.user.generated</packageName>
+                            <sources>
```
[the remark this answers is not in the document]
Comment (kevdoran) [reply nifi-registry-security/pom.xml]: Good find, will update that plugin version.

File: nifi-registry-framework/src/main/java/org/apache/nifi/registry/service/RegistryService.java [hunk 8]
```diff
@@ -261,6 +262,29 @@ public Bucket deleteBucket(final String bucketIdentifier) {
...
     }
 
+    public List<BucketItem> getBucketItems(final QueryParameters queryParameters, final Set<String> bucketIdentifiers) {
+        if (bucketIdentifiers == null || bucketIdentifiers.isEmpty()) {
+            throw new IllegalArgumentException("Bucket Identifiers cannot be null or empty");
+        }
+
+        readLock.lock();
+        try {
+            final Set<BucketEntity> filterBuckets = bucketIdentifiers.stream()
```
[the remark this answers is not in the document]
Comment (kevdoran) [reply nifi-registry-framework/src/main/java/org/apache/nifi/registry/service/RegistryService.java]: Yeah I agree that would be a better approach. Will make that change, thanks.

File: nifi-registry-framework/src/main/java/org/apache/nifi/registry/authorization/StandardAuthorizerFactory.java [hunk 9]
```diff
@@ -0,0 +1,791 @@
...
+        // method injection
+        performMethodInjection(instance, authorizerClass);
+
+        // field injection
+        performFieldInjection(instance, authorizerClass);
+
+        // call post construction lifecycle event
+        instance.initialize(new StandardAuthorizerInitializationContext(identifier, this, this, this));
+
+//        if (StringUtils.isNotEmpty(classpathResources)) {
```
[the remark this answers is not in the document]
Comment (kevdoran) [reply nifi-registry-framework/src/main/java/org/apache/nifi/registry/authorization/StandardAuthorizerFactory.java]: Yeah hadn't gotten that far :)

Its a good point. I'll capture the outstanding work in a JIRA ticket and some in-line TODOs.

File: nifi-registry-framework/src/main/java/org/apache/nifi/registry/service/AuthorizationService.java [hunk 10]
```diff
@@ -0,0 +1,696 @@
...
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    public AccessPolicy updateAccessPolicy(AccessPolicy accessPolicy) {
+        verifyAccessPolicyProviderIsConfigurable();
+
+        // Don't allow changing action or resource of existing policy (should only be adding/removing users/groups)
+        org.apache.nifi.registry.authorization.AccessPolicy currentAccessPolicy =
```
[the remark this answers is not in the document]
Comment (kevdoran) [reply nifi-registry-framework/src/main/java/org/apache/nifi/registry/service/AuthorizationService.java]: The idea here is that no user can change the accessPolicy identifier, resource, or action fields on update. They can only change groups or users that have an access policy. To change resource or action, they would have to delete and create a new access policy.

So in your example, user1 and user2 both get the initial version, but those should be guaranteed to match. Unless I'm making a mistake in my logic...

That said, I think I'll move it to inside the write lock anyway, as it won't hurt and would protect  this code block from future changes that break the expectation I had.

**Comment by kevdoran** [comment]
Thanks a lot for the review @bbende! Great suggestions. Will make those changes and update.

**Comment by scottyaslan** [comment]
@kevdoran With this PR when I try to GET http://localhost:8080/nifi-registry-api/items I am getting a "400 Bad Request: Bucket Identifiers cannot be null or empty"...

**Comment by kevdoran** [comment]
Good catch @scottyaslan - will update with a fix.

**Comment by kevdoran** [comment]
@bbende I restructured the modules based on your suggestion. I now have the following modules:

```
nifi-registry-security                 // just a container module, packaged as pom
  +--nifi-registry-security-api        // gets placed in lib dir for others to implement against
  +--nifi-registry-security-api-impl   // NiFi Registry's implementation of security-api
  +--nifi-registry-security-framework  // extends registry-framework to add security
```

Let me know if you agree with this or if it is overkill for what we need. Technically `nifi-registry-security-api-impl` and `nifi-registry-security-framework` can be combined (technically, they could even be part of `nifi-registry-framework`). I can consolidate some of these if you'd rather keep a flatter, more condensed module structure. Let me know.

**Comment by bbende** [comment]
Thanks for the update! Everything is looking good.

I'm thinking maybe we can fold `nifi-registry-security-framework` into `nifi-registry-framework`, and then move `nifi-registry-security-api-impl` and `nifi-registry-security-api` to top-level modules, just to keep it a little flatter.

**Comment by bbende** [comment]
+1 Latest update looks good, going to merge shortly, thanks!

**Pull request closed without merging**